### PR TITLE
LibThread

### PR DESCRIPTION
**Pull request by bugaevc**
This is a new C++ threading library. You can create threads like this:

```c++
Thread::Thread thread = [yay, captures] {
    do_stuff();
    return 0;
};
thread.start();
```

and there's also an utility to do some CPU-intensive work on a (lazily-spawned) background thread:

```c++
auto action = Thread::BackgroundAction<Result>::create([also, captures] {
    do_some(work);
    return the_result;
});
action.set_on_complete([captures, here, as, well](Result the_result) {
    // This runs on the main thread!
    make_use_of(the_result);
});
```

Both `Thread::Thread` and `Thread::BackgroundAction` are `CObject`s, so you can look at them using the Inspector.

I've ported everything that was using either `CThread` or raw `create_thread()` to LibThread, and removed `CThread`.

### CODE REVIEW

File: Libraries/LibThread/BackgroundAction.h [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+public:
+    static NonnullRefPtr<BackgroundAction<Result>> create(Function<Result()> action)
+    {
+        return adopt(*new BackgroundAction(move(action)));
+    }
+
+    ~BackgroundAction() {}
+
+    void set_on_complete(Function<void(Result)> callback)
```
[the remark this answers is not in the document]
Comment (bugaevc) [reply Libraries/LibThread/BackgroundAction.h]: Yes, I did just that originally. But this function is more than a simple setter, it has to sync with the action itself executing concurrently, and run the callback in place if the action has already completed.

An alternative would be not to have actions autostart when constructed, so you'd set the callback first and then call `.start()` explicitly.

[the remark this answers is not in the document]
Comment (awesomekling) [reply Libraries/LibThread/BackgroundAction.h]: Okay, then let's just change the name so it doesn't sound like a straightforward setter. :)

File: Servers/WindowServer/WSCPUMonitor.cpp [hunk 2]
```diff
@@ -9,26 +9,26 @@
 #include <unistd.h>
 
 WSCPUMonitor::WSCPUMonitor()
-{
-    create_thread([](void* context) -> int {
-        auto& monitor = *(WSCPUMonitor*)context;
+    : m_thread([this] {
```
[the remark this answers is not in the document]
Comment (bugaevc) [reply Servers/WindowServer/WSCPUMonitor.cpp]: Kind of, yeah. I guess I should make this a member function, and pass a
```c++
m_thread([this] { member_func(); return 0; })
```
in here — or is there a better way?

**Comment by bugaevc** [comment]
Tweaks following an IRC discussion; now using `LibThread::` as a namespace instead of just `Thread::`, passing two lambdas directly into `BackgroundAction::create()`, and with `CLock` moved to `LibThread::Lock`

**Comment by awesomekling** [comment]
Okay! :)

**Pull request merged**